types/single-spa-react.d.ts

```diff
@@ -1,21 +1,23 @@
 import * as React from "react";
-import * as ReactDOM from "react-dom";
+import * as ReactDOMClient from "react-dom/client";
 import { AppProps, CustomProps, LifeCycleFn } from "single-spa";
 
 export const SingleSpaContext: React.Context<CustomProps & AppProps>;
 
+export interface ReactAppOrParcel<ExtraProps> {
+  bootstrap: LifeCycleFn<ExtraProps>;
+  mount: LifeCycleFn<ExtraProps>;
+  unmount: LifeCycleFn<ExtraProps>;
+  update?: LifeCycleFn<ExtraProps>;
+}
+
+type LoadRootComponentProps = any;
+
 export interface SingleSpaReactOpts<RootComponentProps> {
   React: typeof React;
-  ReactDOM: typeof ReactDOM;
-  rootComponent?:
-    | React.ComponentClass<RootComponentProps, any>
-    | React.FunctionComponent<RootComponentProps>;
-  loadRootComponent?: (
-    props: RootComponentProps
-  ) => Promise<
-    | React.ComponentClass<RootComponentProps, any>
-    | React.FunctionComponent<RootComponentProps>
-  >;
+  ReactDOMClient: typeof ReactDOMClient;
+  rootComponent?: React.ElementType<RootComponentProps>;
+  loadRootComponent?: (props?: any) => Promise<React.ElementType<any>>;
```

Comment:
I encountered a strange TypeScript-ism with the previous typings for this. Because they are using the same type, if the outer function doesn't have props, the type is inferred as `never` which then gets collapsed/passed/inferred down (idk what to call this behavior) to the props being passed to the component even though you might not need them. More concretely, it means you cannot have inner props defined on the function that the Promise resolves with if the outer props aren't also defined.

```js
loadRootComponent: () => Promise.resolve(() => <div>hi</div>)) // OK
loadRootComponent: (outerProps) => Promise.resolve((innerProps) => <div>hi</div>)) // OK
loadRootComponent: (outerProps) => Promise.resolve(() => <div>hi</div>)) // OK
loadRootComponent: () => Promise.resolve((innerProps) => <div>hi</div>)) // ERROR
```

Loosening the type here meant that you can now use that last combination without error. This wasn't worth the time I put into it but its done now.